packages/legend-application/src/stores/storage/StorageService.ts

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright (c) 2020-present, Goldman Sachs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  type PlainObject,
+  SerializationFactory,
+  usingModelSchema,
+} from '@finos/legend-shared';
+import { createModelSchema, list, primitive } from 'serializr';
+import type { GenericLegendApplicationStore } from '../ApplicationStore.js';
+import { LocalStorage } from './LocalStorage.js';
+
+const APPLICATION_SETTINGS_STORAGE_KEY = 'application-settings-storage';
+
+export class StorageItem {
+  key!: string;
+  value!: string;
+
+  constructor(key: string, value: string) {
+    this.key = key;
+    this.value = value;
+  }
+
+  static create(key: string, value: string): StorageItem {
+    return new StorageItem(key, value);
+  }
+}
+
+const storageItemModelSchema = createModelSchema(StorageItem, {
+  key: primitive(),
+  value: primitive(),
+});
+
+export class StorageStore {
+  items: StorageItem[] = [];
+
+  static readonly serialization = new SerializationFactory(
+    createModelSchema(StorageStore, {
+      items: list(usingModelSchema(storageItemModelSchema)),
+    }),
+  );
+
+  static serialize(value: StorageStore): PlainObject<StorageStore> {
+    return StorageStore.serialization.toJson(value);
+  }
+}
+
+export class StorageService {
+  readonly applicationStore: GenericLegendApplicationStore;
+  readonly localStorage: LocalStorage;
+  readonly storageKey: string;
```

Comment:
I think what we should do is expose several stores, e.g. `settingsStore!: StorageStore;`, for each `StorageStore`, we have:

```ts
type StoredValue = object | string | number | boolean; 

// NOTE: in the future, we could potentially support a schema for the store so as to have better type-checking
class StorageStore {
  readonly key: string = <SOME CONSTANT>;
  readonly storageService!: StorageService;
  private value!: Record<PropertyKey, StoredValue>;

  getValue(key: string): StoredValue;
  getNumberValue(key: string): number; // does validation using `isNumber`
  getBooleanValue(key: string): string; // does validation using `isString`
  getBooleanValue(key: string): boolean; // does validation using `osBoolean`
  getObjectValue(key: string): number; // does validation using `isObject`
  persist(key: string, value: StoredValue);
}

```

Reply:
every time we persist, we update the value here and the value of the storage